enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java

```diff
@@ -30,87 +32,159 @@
+     *             "hash": BYTE[],
+     *             "salt": BYTE[]
+     *         }
+     *     },
+     *     ...
+     * }
+     */
     public static UsersMetaData fromXContent(XContentParser parser) throws IOException {
-        List<String> users = new ArrayList<>();
-        if (parser.nextToken() == XContentParser.Token.FIELD_NAME && parser.currentName().equals("users")) {
-            if (parser.nextToken() == XContentParser.Token.START_ARRAY) {
-                XContentParser.Token token;
-                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY && token != null) {
-                    users.add(parser.text());
-                }
-            }
-            if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
-                // each custom metadata is packed inside an object.
-                // each custom must move the parser to the end otherwise possible following customs won't be read
-                throw new ElasticsearchParseException("failed to parse users, expected an object token at the end");
+        Map<String, UserAttributes> users = new HashMap<>();
+
+        while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+            String userName = parser.currentName();
+            if (parser.nextToken() == XContentParser.Token.START_OBJECT) {  // user
+                users.put(userName, attributesFromXContent(parser));
             }
         }
         return new UsersMetaData(users);
     }
 
+    private static UserAttributes attributesFromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token currentToken;
+        int iterations = 0;
+        byte[] hash = new byte[0];
+        byte[] salt = new byte[0];
+        boolean hasPassword = false;
+
+        while ((currentToken = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (currentToken == XContentParser.Token.FIELD_NAME) {
+                while ((currentToken = parser.nextToken()) != XContentParser.Token.END_OBJECT) { // secure_hash
+                    hasPassword = true;
+                    if (currentToken == XContentParser.Token.FIELD_NAME) {
+                        String currentFieldName = parser.currentName();
+                        currentToken = parser.nextToken();
+                        switch (currentFieldName) {
+                            case "iterations":
+                                if (currentToken != XContentParser.Token.VALUE_NUMBER) {
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'iterations' value is not a number [{}]", currentToken);
+                                }
+                                iterations = parser.intValue();
+                                break;
+                            case "hash":
+                                if (currentToken == XContentParser.Token.VALUE_NULL) {
```

Comment:
@seut do you have a proper solution what Token type for `byte[]` could be used instead of a `VALUE_NULL` check ?